src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1064,304 +907,250 @@ internal TransitionRegex<S> CreateDerivative()
+        /// <param name="eager">whether the derivative should only consider paths that backtracking would take, true by default</param>
+        /// <returns>true if an unconditionally nullable transition has been generated, which means that backtracking would not consider any further paths</returns>
+        private bool AddTransitions(S elem, uint context, List<(SymbolicRegexNode<S>, DerivativeEffect[]?)> transitions,
+            List<SymbolicRegexNode<S>> continuation, Stack<DerivativeEffect>? effects, bool eager = true)
+        {
+            // Helper function for concatenating a head node and a list of continuation nodes. The continuation nodes
+            // are added in reverse order and the function below uses the list as a stack, so the nodes added to the
+            // stack first end up at the tail of the concatenation.
+            static SymbolicRegexNode<S> BuildLeaf(SymbolicRegexNode<S> head, List<SymbolicRegexNode<S>> continuation)
             {
-                return transition;
+                SymbolicRegexNode<S> leaf = head._builder.Epsilon;
+                for (int i = 0; i < continuation.Count; ++i)
+                {
+                    leaf = head._builder.CreateConcat(continuation[i], leaf);
+                }
+                return head._builder.CreateConcat(head, leaf);
             }
 
+            // Nothing and epsilon can't consume a character so they generate no transition
             if (IsNothing || IsEpsilon)
             {
-                transition = TransitionRegex<S>.Leaf(_builder._nothing);
-                return transition;
+                return false;
             }
 
+            // For both .* and .+ the derivative is .* for any character
             if (IsAnyStar || IsAnyPlus)
             {
-                transition = TransitionRegex<S>.Leaf(_builder._anyStar);
-                return transition;
+                SymbolicRegexNode<S> leaf = BuildLeaf(_builder._anyStar, continuation);
+                transitions.Add((leaf, effects?.ToArray()));
+                // Signal early exit if the leaf is unconditionally nullable
+                return leaf.IsNullable;
             }
 
             if (!StackHelper.TryEnsureSufficientExecutionStack())
             {
-                return StackHelper.CallOnEmptyStack(CreateDerivativeWithEffects, eager);
+                return StackHelper.CallOnEmptyStack(AddTransitions, elem, context, transitions, continuation, effects, eager);
             }
 
             switch (_kind)
             {
                 case SymbolicRegexNodeKind.Singleton:
                     Debug.Assert(_set is not null);
-                    transition = TransitionRegex<S>.Conditional(_set, TransitionRegex<S>.Leaf(_builder.Epsilon), TransitionRegex<S>.Leaf(_builder._nothing));
+                    // The following check assumes that either (1) the element and all predicates are minterms, in which case
```

Comment:
Ah yes, maybe it would be clearer to just say that "the element and predicate are minterms" (in which case all predicates in the regex are also minterms).